### PR TITLE
FA-19 Dialog for adding a layer

### PR DESCRIPTION
**Pull request by brosscode**
Adds a Dialog to add a new layer. Includes limitations on name of layer and will persist on rotation.

### CODE REVIEW

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt [hunk 1]
```diff
@@ -36,11 +37,31 @@ class CustomColorViewModel : ViewModel() {
         currentBitmap = b
     }
 
+    var isDialogOpen = false
+    var layerName: String? = null
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: `currentUserEnteredLayerName`?

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt [hunk 2]
```diff
@@ -36,11 +37,31 @@ class CustomColorViewModel : ViewModel() {
...
+    private var layerNumber: Int = 1
+
+    fun makeLayerNameHint(s: String): String {
+        return "$s $layerNumber"
+    }
+
+    fun addLayer(s: String) {
+        if (s.length > NAME_LENGTH)
+            if (s == ("Layer $layerNumber")) {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: This should be in a string resource. Look up formattable string resources

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt [hunk 3]
```diff
@@ -36,11 +37,31 @@ class CustomColorViewModel : ViewModel() {
         currentBitmap = b
     }
 
+    var isDialogOpen = false
+    var layerName: String? = null
+    private var layerNumber: Int = 1
+
+    fun makeLayerNameHint(s: String): String {
+        return "$s $layerNumber"
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: Resource

Comment (ForzenAcademy) [reply app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: To clarify, the reason this should be a resource also is that your order won't be the same in every language

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt [hunk 4]
```diff
@@ -36,11 +37,31 @@ class CustomColorViewModel : ViewModel() {
...
+    }
+
+    fun addLayer(s: String) {
+        if (s.length > NAME_LENGTH)
+            if (s == ("Layer $layerNumber")) {
+                layerNumber += 1
+            }
+    }
+
+
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt:86]: Extra line

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt [hunk 5]
```diff
@@ -17,11 +25,63 @@ class MainActivity : AppCompatActivity() {
...
     }
+
+
+    @SuppressLint("ClickableViewAccessibility")
+    private fun colorPickerDialog() {
+        viewModel.isDialogOpen = true
+        val inflater = LayoutInflater.from(this@MainActivity)
+        val view = inflater.inflate(R.layout.add_layer_dialog, null)
+        val dialog = AlertDialog.Builder(this).apply {
+            setTitle("Add Layer")
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/MainActivity.kt]: String resource

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt [hunk 6]
```diff
@@ -17,11 +25,63 @@ class MainActivity : AppCompatActivity() {
...
+            setView(view)
+            val errorView = view.findViewById<TextView>(R.id.errorText)
+            val layerNameInput = view.findViewById<EditText>(R.id.layerNameField)
+            layerNameInput.hint = viewModel.makeLayerNameHint(context.getString(R.string.AddLayer))
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/MainActivity.kt]: resources should be lower case and snake case

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt [hunk 7]
```diff
@@ -17,11 +25,63 @@ class MainActivity : AppCompatActivity() {
...
+            if (viewModel.isDialogOpen && viewModel.layerName != null) {
+                layerNameInput.setText(viewModel.layerName)
+                if (layerNameInput.text.toString().length > 16) {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/MainActivity.kt]: You have the constant for this

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt [hunk 8]
```diff
@@ -17,11 +25,63 @@ class MainActivity : AppCompatActivity() {
...
+                    errorView.visibility = View.VISIBLE
+                }
+            }
+            layerNameInput.addTextChangedListener {
+                viewModel.layerName = it.toString()
+                if (it.toString().length > 16) {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/MainActivity.kt]: Constant

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt [hunk 9]
```diff
@@ -17,11 +25,63 @@ class MainActivity : AppCompatActivity() {
...
+                    errorView.visibility = View.VISIBLE
+                }
+            }
+            setPositiveButton("Ok") { _, _ ->
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/MainActivity.kt]: String resource

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt [hunk 10]
```diff
@@ -17,11 +25,63 @@ class MainActivity : AppCompatActivity() {
...
+                if (layerName.isEmpty()) {
+                    layerName = viewModel.makeLayerNameHint(context.getString(R.string.AddLayer))
+                    viewModel.addLayer(layerName)
+                } else {
+                    viewModel.addLayer(layerName)
+                }
+                viewModel.isDialogOpen = false
+                viewModel.layerName = null
+            }
+            setNegativeButton("Cancel") { _, _ ->
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/MainActivity.kt]: String resource

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt (hunk 1 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: Changed to name provided

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt (hunk 2 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: Created a formatted string in the Strings XML.

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt (hunk 3 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: Deleted this function as I am now using a formatted string xml. There is a lot of consideration that has to go into localization for strings, is what I learned.

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt (hunk 5 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/MainActivity.kt]: Now uses a string resource for dialog title

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt (hunk 4 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt:86]: Line removed

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt (hunk 6 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/MainActivity.kt]: adjusted to follow standard naming conventions

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt (hunk 7 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/MainActivity.kt]: Now uses a constant

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt (hunk 8 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/MainActivity.kt]: Now uses a constant

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt (hunk 9 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/MainActivity.kt]: Added String Resource

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt (hunk 10 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/MainActivity.kt]: Added String Resource

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt [hunk 11]
```diff
@@ -36,11 +38,28 @@ class CustomColorViewModel : ViewModel() {
         currentBitmap = b
     }
 
+    var isDialogOpen = false
+    var currentUserEnteredLayerName: String? = null
+    var layerNumber: Int = 1
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: Constant

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt [hunk 12]
```diff
@@ -36,11 +38,28 @@ class CustomColorViewModel : ViewModel() {
...
+
+    fun addLayer(s: String, b: Boolean) {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: What are these arguments?

Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: Kdoc

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt [hunk 13]
```diff
@@ -36,11 +38,28 @@ class CustomColorViewModel : ViewModel() {
...
+        if (s.length < NAME_LENGTH) {
+            if (b) {
+                layerNumber += 1
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: Is this going to work if layers are deleted? Or renamed? Or if the file is saved and reloaded?

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt [hunk 14]
```diff
@@ -36,11 +38,28 @@ class CustomColorViewModel : ViewModel() {
...
     private fun updateViews() {
         onUpdate?.invoke(
             ViewState(
                 primaryColor, currentBitmap
             )
         )
     }
+
+    companion object InputLimitations {
+        const val NAME_LENGTH = 16
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: `MAX_`?

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt [hunk 15]
```diff
@@ -17,11 +25,83 @@ class MainActivity : AppCompatActivity() {
         viewModel.currentBitmap?.let { drawableArea.setBitmap(it) }
         val colorSelector = findViewById<PrimaryColorCircleView>(R.id.colorCircle)
         colorSelector.onChangeColor = { viewModel.cyclePrimaryColor() }
+        findViewById<Button>(R.id.add_layer).setOnClickListener {
+            colorPickerDialog()
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/MainActivity.kt]: `openColorPickerDialog`?

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt [hunk 16]
```diff
@@ -17,11 +25,83 @@ class MainActivity : AppCompatActivity() {
...
+        }
+
+        if (viewModel.isDialogOpen) {
+            colorPickerDialog()
+        }
+
+        viewModel.onAddLayerSuccess = {
+            Toast.makeText(this, getString(R.string.addLayerSuccess) + it, Toast.LENGTH_SHORT)
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/MainActivity.kt]: This should be a formatted string. Look up how to do this. You're going to get string with arguments to format it

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt [hunk 17]
```diff
@@ -17,11 +25,83 @@ class MainActivity : AppCompatActivity() {
...
+        val dialog = AlertDialog.Builder(this).apply {
+            setTitle(context.getString(R.string.addLayerDialogTitle))
+            setView(view)
+            val errorView = view.findViewById<TextView>(R.id.errorText)
+            val layerNameInput = view.findViewById<EditText>(R.id.layerNameField)
+            layerNameInput.hint =
+                context.getString(R.string.addLayerDefaultName, viewModel.layerNumber)
+            if (viewModel.isDialogOpen && viewModel.currentUserEnteredLayerName != null) {
+                layerNameInput.setText(viewModel.currentUserEnteredLayerName)
+                if (layerNameInput.text.toString().length > NAME_LENGTH) {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/MainActivity.kt]: You can reduce this.
`errorView.isVisible = layerNameInput.text.toString().length > NAME_LENGTH`

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt [hunk 18]
```diff
@@ -17,11 +25,83 @@ class MainActivity : AppCompatActivity() {
...
+                    errorView.visibility = View.VISIBLE
+                } else {
+                    errorView.visibility = View.INVISIBLE
+                }
+            }
+            layerNameInput.addTextChangedListener {
+                viewModel.currentUserEnteredLayerName = it.toString()
+                if (it.toString().length > NAME_LENGTH) {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/MainActivity.kt]: You can simplify this too

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt [hunk 19]
```diff
@@ -17,11 +25,83 @@ class MainActivity : AppCompatActivity() {
...
+                    errorView.visibility = View.VISIBLE
+                } else {
+                    errorView.visibility = View.INVISIBLE
+                }
+            }
+            setPositiveButton(getString(R.string.addLayerPositiveButton)) { _, _ ->
+                var layerName: String = layerNameInput.text.toString()
+                if (layerName == context.getString(
+                        R.string.addLayerDefaultName,
```
Comment (ForzenAcademy) [inline app/src/main/java/com/example/drawingactivity/MainActivity.kt]: Ugly. Reformat

**Review by ForzenAcademy — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt (hunk 11 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: Added a constant for this

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt (hunk 14 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: Renamed to NAME_LENGTH_MAX

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt (hunk 15 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/MainActivity.kt]: Renamed to suggestion

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt (hunk 19 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/MainActivity.kt]: Reformatted to a when block

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt (hunk 17 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/MainActivity.kt]: reduced to given statement

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt (hunk 18 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/MainActivity.kt]: Simplified

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt (hunk 12 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: removed the boolean, renamed string to LayerName

Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: added a kdoc

File: app/src/main/java/com/example/drawingactivity/MainActivity.kt (hunk 16 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/MainActivity.kt]: Reformatted String

File: app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt (hunk 13 above)
Comment (brosscode) [reply app/src/main/java/com/example/drawingactivity/CustomColorViewModel.kt]: Implemented more robust logic for adding layers based on what names are in use and which are not.

**Review by ForzenAcademy — APPROVED** [review]

**Pull request merged**